Map `bioformats` images with tiles in x-y dimension in `dask`

NHPatterson opened the pull request:
## Description
The recent bioformats reader adds fantastic support for all things bioformats, but `dask` chunks were mapped as planes which is less than ideal when working with whole slide images or and image with very large x-y planes. Fortunately, most of the logic to support tiling was already present so I added some code to split the y and x axis up by tiles of arbitrary dimension and map those with `da.map_blocks`. 

There are two new `kwargs` in `BioFile`:

`dask_tiles` -> use dask tiles, bool
`tile_size` -> optional tuple of tile size in y and x, otherwise defaults to recommended by bioformats with [`getOptimalTileHeight/Width'](https://downloads.openmicroscopy.org/bio-formats/5.4.1/api/loci/formats/IFormatReader.html#getOptimalTileHeight--)

I've run some small test locally on WSIs from 2 formats with success, but the current test suite does not include any images where such tiling would be appropriate, so I'm keeping this as draft to see how best to progress and get feedback on how to proceed.

### quick memory profile on local data
```bash
python -m memory_profiler bf-memtest.py

```
```python
import os
os.environ["BIOFORMATS_MEMORY"] = "8g"
import numpy as np
from aicsimageio.readers.bioformats_reader import BioFile

@profile
def mem_consumption():
    bf = BioFile("./data/VAN0032-RK-3-10-PAS_IMS.scn",series=3, dask_tiles=True)
    partial_by_tile = bf.to_dask()[:, :, :, 1024:2048, 1024:4096].compute()
    bf = BioFile("./data/VAN0032-RK-3-10-PAS_IMS.scn",series=3, dask_tiles=False)
    partial_by_plane = bf.to_dask()[:, :, :, 1024:2048, 1024:4096].compute()
    assert np.array_equal(partial_by_tile, partial_by_plane)


if __name__ == '__main__':
    mem_consumption()
```
**Terminal output**
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     7  220.094 MiB -1415.418 MiB           2   @profile
     8                                         def mem_consumption(use_tiles=False):
     9  268.086 MiB  172.887 MiB           2       bf = BioFile("./data/VAN0032-RK-3-10-PAS_IMS.scn",series=3, dask_tiles=True)
    10  298.879 MiB   51.410 MiB           2       partial_by_tile = bf.to_dask()[:, :, :, 1024:2048, 1024:4096].compute()
    11  308.559 MiB    4.121 MiB           2       bf = BioFile("./data/VAN0032-RK-3-10-PAS_IMS.scn",series=3, dask_tiles=False)
    12 3475.008 MiB 4624.738 MiB           2       partial_by_plane = bf.to_dask()[:, :, :, 1024:2048, 1024:4096].compute()
    13 3475.008 MiB    0.301 MiB           2       assert np.array_equal(partial_by_tile, partial_by_plane)

```
^^ massive decrease in memory consumption to read a partial area, arrays are equal. 

## Pull request recommendations:
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

NHPatterson commented:
> thanks @NHPatterson ... indeed, XY chunking was left as a feature to be added later (and also, I didn't have any good tiled data to test 😄 ). Thanks for picking up the torch!

Happy to do it!

> what would you think about making this an argument to the `to_dask` method, rather than a parameter to `__init__`? 

That makes sense, the same thought occurred to me, I'll make that change, but I have to plead ignorance on some aspects here, how would users access this from the `AICSImage`? 

> Also, do you think that `tile_size` and `dask_tiles` could be reduced to a single parameter? For instance, if `tile_size` is `None`, it's the same as `dask_tiles == False`?

Are you suggesting to have `tile_size` as a single argument? My concern is if a user wants to use the suggested tile size from bioformats, this would be difficult to find prior. Currently, setting `dask_tiles` to True and `tile_size` to `None` will use tile sizes suggested by bf (probably good defaults as they try to map file internal tiling). It could be `tile_size` == `True` uses defaults but the name and argument type seem confusing. 

> @JacksonMaxfield can help out with getting a suitable test file into the test data

There are some on the openslide site that may be appropriate: http://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/

NHPatterson commented:
> > That makes sense, the same thought occurred to me, I'll make that change, but I have to plead ignorance on some aspects here, how would users access this from the AICSImage?
> 
> I have to plead some ignorance too! 😂 i think we'll need some input from @JacksonMaxfield or @toloudis ... I _think_ the solution here is to add a `MosaicTile` dimension to the xarray data coming from the Bioformats reader. that would then get recognized here:
> 
> https://github.com/AllenCellModeling/aicsimageio/blob/0026688ee078997799329d843d898720d54dd41e/aicsimageio/aics_image.py#L373-L378
> 
> and assuming the AICSImage was created with `reconstruct_mosaic==True` (the default), then AICSImage would take care of it.
> 
> In other words, the main way that the lower level readers communicate with AICSImage is by properly filling in those xarray labels, so I guess what we really have to here is update the dims/coords in the BioformatsReader._to_xarray method:
> 
> https://github.com/AllenCellModeling/aicsimageio/blob/0026688ee078997799329d843d898720d54dd41e/aicsimageio/readers/bioformats_reader.py#L176-L179
> 
> (and that method could provide the params to `to_dask()` mentioned above)

Hmm.. For the `AICSImage`, I am still unsure about the mosaic aspect. it seems that `AICSImage` expects an M dimension at index 0, then it uses this to put all the tiles together, whereas this is currently providing an already stitched image (I don't think there is really a mosaic dimension in reality, like you might see with a zeiss CZI where the API allows you to read in tile by tile as parameter), just with the chunking defined by the internal data tiles. Could use some insight from @JacksonMaxfield and @toloudis TBH on what is best to do..

JacksonMaxfield commented:
Hello all! This is an exciting PR for sure and thanks for getting it started @NHPatterson 

I am going to respond to everything I can see but if I miss something please just re-call-it-out :sweat_smile:

> There are some on the openslide site that may be appropriate: http://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/

@NHPatterson can you tell me the filename (or two) from that set that you would like me to add to the test resources? Happy to use any of the ones in the linked website but don't know which would be best for testing.

>> I have to plead some ignorance too! joy i think we'll need some input from @JacksonMaxfield or @toloudis ... I think the solution here is to add a MosaicTile dimension to the xarray data coming from the Bioformats reader. that would then get recognized here: ...

> Hmm.. For the AICSImage, I am still unsure about the mosaic aspect. it seems that AICSImage expects an M dimension at index 0 and each tile is iterated by M. It then uses this to put all the tiles together, whereas rearranging the chunking in this PR provides planes already stitched image. Could use some insight from @JacksonMaxfield and @toloudis TBH on what is best to do.

Yes I was going to say that that if these are mosaic images, then try to use the `M` dimension and place your code into the stitch tile function however from your explanation here these are _pre-stitched_ images. Hmmm. Interesting. There is a part of me that wants to say "fake it" because ultimately the desired behavior is to get some XY chunk out at a time. But I understand the distinction is important.

As a reminder, `AICSImage` passes kwargs down to the readers. So if the user did `AICSImage("some-file.xyz", reader=BioformatsReader, dask_tiles=True)` it would get passed down properly (I think.... although this was being discussed in #321).

Sorry I have been thinking out loud here but all summed I think we just need to come to an agreement on if this should fake an M-index / M dimension per tile. I can see arguments on both sides.

Pros of faking an M-index:
1. XY chunking behavior keeps in line with the rest of the AICSImage API about "when to expect chunked XY"
2. extra metadata (tile position, tile size) functions become available (this is super minor imo)
3. we could somewhat just turn this on for any image larger than some X dim size? (or is there metadata in OME that can inform of this decision?)

Cons of faking an M-index:
1. adds complexity to the codebase
2. non-intuitive to the user on how the data is actually stored in the file

If there is metadata in the OME blob that will tell us if the image is stitched or not, I think I side on the "do this by default and fake the m-index in that case" side of things. If not, I think the current implementation with the parameter in the `__init__` is the best method for this behavior because of how parameters can be passed down to the `Reader` from `AICSImage`.

(I am sure you all can think of more pros and cons, so please do add to those lists)

---

@psobolewskiPhD, @tlambert03's explanation is entirely correct. Each reader has it's own tile stitcher. As a general statement, I think basically every reader in this library should be switched over to using `map_blocks` to do file reading and tile stitching. For example, you can find [basically this exact same loop across most of the readers](https://github.com/AllenCellModeling/aicsimageio/blob/main/aicsimageio/readers/default_reader.py#L436) to do basic file reading / stacking. I am 99% sure everywhere that is used should be replaced with a single `map_blocks` call. Lot's to do and never enough time.

toloudis commented:
> > @JacksonMaxfield can correct me if I'm wrong, but the high level intention of mosaic is for images where whole XY "planes" are really tiles meant to be stitched back together. So I'm not sure it fits this use case. I think if we overloaded its usage to handle dask sub-xy chunks, then we might risk losing some performance and gain code complexity?
> > (I'm generally against overloading M in this way)
> 
> I think that's fair, however if these images are majority "pre-stitched" mosaics (and if there is metadata that indicates they are) then I can't help but feel that we would just be helping the user if we maintain a standard behavior especially if it helps in their processing of large images.
> 
> And I would slightly disagree that the intention of the mosaic API was to stitch planes together, from my perspective, the mosaic API was that _and more importantly_ to allow the user to "fit" their data in-memory (or in the dask sense, in a single array)

I'm just saying M was never intended to be used to split up planes that weren't split in the original files.  It came from the way CZI stores mosaic tiles, one per subblock where a subblock is logically equivalent to a tiff plane.

JacksonMaxfield commented:
Meta-comment, when will GitHub add threading to comments? We desparately need it.

---

> I'm just saying M was never intended to be used to split up planes that weren't split in the original files. It came from the way CZI stores mosaic tiles, one per subblock where a subblock is logically equivalent to a tiff plane.

Fair. I would say then we should allow this as a Reader init parameter?

NHPatterson commented:
Thanks for the breakdown on this, seems there is still some debate.

For the time being, I think I got this to work using a "fake" M dimension. I added methods that produce a `mosaic_xarray_dask_data`, but this method really doesn't do anything different but load the "tiled" version of the dask image. Unfortunately, the normal xarray check with the fake M result in a dask array with 6 dimensions (always with length 1 on the M dimension) rather than the 5 normal for bioformats. I added a check to drop this in `AICSImage`, but not sure if this is too much of a hack.

```python
from aicsimageio import AICSImage
im = AICSImage("T:/CMU-1-Small-Region.svs", reconstruct_mosaic=True)
im.set_scene(0)
print(im.dask_data)
```
```python
dask.array<transpose, shape=(1, 1, 1, 2967, 2220, 3), dtype=uint8, chunksize=(1, 1, 1, 240, 240, 3), chunktype=numpy.ndarray>
```
```python
im = AICSImage("T:/CMU-1-Small-Region.svs", reconstruct_mosaic=False)
im.set_scene(0)
print(im.dask_data)
```
```python
dask.array<getitem, shape=(1, 1, 1, 2967, 2220, 3), dtype=uint8, chunksize=(1, 1, 1, 2967, 2220, 3), chunktype=numpy.ndarray>
```

toloudis commented:
> Thanks for the breakdown on this, seems there is still some debate.
> 
> For the time being, I think I got this to work using a "fake" M dimension. I added methods that produce a `mosaic_xarray_dask_data`, but this method really doesn't do anything different but load the "tiled" version of the dask image. Unfortunately, the normal xarray check with the fake M result in a dask array with 6 dimensions (always with length 1 on the M dimension) rather than the 5 normal for bioformats. I added a check to drop this in `AICSImage`, but not sure if this is too much of a hack.
> 
> ```python
> from aicsimageio import AICSImage
> im = AICSImage("T:/CMU-1-Small-Region.svs", reconstruct_mosaic=True)
> im.set_scene(0)
> print(im.dask_data)
> ```
> 
> ```python
> dask.array<transpose, shape=(1, 1, 1, 2967, 2220, 3), dtype=uint8, chunksize=(1, 1, 1, 240, 240, 3), chunktype=numpy.ndarray>
> ```
> 
> ```python
> im = AICSImage("T:/CMU-1-Small-Region.svs", reconstruct_mosaic=False)
> im.set_scene(0)
> print(im.dask_data)
> ```
> 
> ```python
> dask.array<getitem, shape=(1, 1, 1, 2967, 2220, 3), dtype=uint8, chunksize=(1, 1, 1, 2967, 2220, 3), chunktype=numpy.ndarray>
> ```

Those results look great to me and my only hesitation is the API part of it.  `reconstruct_mosaic=False` intuitively means to me that I'm getting those 240x240 chunks.  (Where did the 240x240 come from?  I'm still missing whether that is user provided or specified in the file somewhere -- I'll look at the code) I feel like it would be better captured with `im.get_image_dask_data("YXS", chunk_dims=(1,1,1,240,240,3), reconstruct=True)` (pseudo-syntax here)

NHPatterson commented:
> Those results look great to me and my only hesitation is the API part of it. `reconstruct_mosaic=False` intuitively means to me that I'm getting those 240x240 chunks. (Where did the 240x240 come from? I'm still missing whether that is user provided or specified in the file somewhere -- I'll look at the code) I feel like it would be better captured with `im.get_image_dask_data("YXS", chunk_dims=(1,1,1,240,240,3), reconstruct=True)` (pseudo-syntax here)

I agree, this doesn't make sense with the reconstruct_mosaic argument as-is. This is just the I saw to insert reading the XY planes in chunks rather than each XY plane being an chunk in and of itself. In this case, the 240x240 comes from the layout of the tiff file on disk. See from Aperio ImageScope (vendor software):
![image](https://user-images.githubusercontent.com/17855764/137795883-d23ce263-27fd-4253-9305-7b86bf46457f.png)

I think the distinction is that this isn't related to the individual tiles as stored during acquisition, but really how the images are tiled post-acquisition for display. The tile size here is determined by default using [IFormatReader](https://downloads.openmicroscopy.org/bio-formats/5.4.1/api/loci/formats/IFormatReader.html#getOptimalTileHeight--), which for most formats, checks the layout on disk. In fact, these images were probably acquired as continuous-raster stripes, registered, then split into tiles during write.

JacksonMaxfield commented:
@NHPatterson if you have a second want to push the TEST RESOURCES HASH change?

Would love to merge this and release a new version of aicsimageio

NHPatterson commented:
> @NHPatterson if you have a second want to push the TEST RESOURCES HASH change?
> 
> Would love to merge this and release a new version of aicsimageio

Apologies for the delay! It's in
